### ISSUE TITLE
Consumer Group - Self Balanced Consumers 

### ISSUE THREAD

**Issue by wadearnold** [segmentio/kafka-go#5]
First thanks for open sourcing this project. I couldn't agree more with your description in the opening readme. Thank you for brining a more idiomatic go solution to the Kafka/Go community! 

My current project is highly dependent on [Sarama Cluster](github.com/bsm/sarama-cluster) because of its support for Consumer Groups with self balancing. 
- High availability of the Consumer group
- Sharding of load during peak processing times. 
- Ensuring a single offset is read by one and only one consumer. 

My simple experiments show that [kafka-go/consu-go](segmentio/consul-go") support Consumer Groups but fall back to the simple consumer(reader) by requiring the partition to be assigned by the client. 

Is there any effort to support what Jay Kreps calls a [High level Consumer](https://cwiki.apache.org/confluence/display/KAFKA/Consumer+Client+Re-Design) inside of Kafka-go? 

Your reader implementation with context support would be an awesome starting point. I would be happy to contribute to this effort. 

Again thank you for your contribution. If I just can't read the code let me know!

**Comment by thehydroimpulse** [comment]
Consumer groups would definitely be really awesome to support. With one of our recent projects we decided to go with consul as a simple way to do consumer groups but required that 1) we stored our offsets ourselves 2) there's no fancy balancing or anything 3) dependent on consul.

For most use cases it's easier to create a real Kafka consumer group and not have to worry about too much.

**Comment by achille-roussel** [comment]
Supporting _native_ Kafka consumer groups are definitely a must-have, I was planning on looking into it sometimes this weekend maybe.

Design proposals / Pull Requests / Thoughts are welcome!

**Comment by ardan-bkennedy** [comment]
I am working with the example code with kafka running in a docker container from Sarama. I am confused how the consumer is able to always get the latest new messages on every restart. I am not maintaing my own offset position. I am very new to kafka. I just need simple consuming and publishing of messages.

Thank you for this package because the other packages were not written very well and I was scared to use them.

**Comment by achille-roussel** [comment]
@ardan-bkennedy currently kafka-go doesn't support consumer groups, which means you have to manage the offset yourself. We focused on this use case because it was fitting our needs here at Segment, however we also have places where having offsets managed automatically is a requirement for us so we're starting to think about building support for consumer groups into this package.

Unfortunately for now your best option is to go with Sarama if you want to have offsets managed for you.

Stay tuned and thanks for the words of support 👍

**Comment by kevinburke** [comment]
Might be a silly question but what's preventing just adding an extra `GroupID` field or similar to the `ReaderConfig` and then sending that to Kafka on dial?

**Comment by kevinburke** [comment]
Reading through the code some more... we can't send a group id as part of a Fetch request. after establishing the connection, if the configured group id is non-nil we'd need to send a JoinGroup message.

**Comment by achille-roussel** [comment]
Consumer groups are a little more complex to implement than just setting an ID unfortunately. They require a bunch of client side logic to be implemented, I'm starting to have more and more use cases that could leverage consumer groups so I'll probably end up implementing it eventually. If you want to give it a try let me know so we don't duplicate efforts.

**Comment by kevinburke** [comment]
Oh interesting, OK. No I don't have any immediate plans to work on this

**Comment by harlow** [comment]
Wondering if there is an incremental approach that could be made here? Maybe this is too simplistic thinking, but we could break this into two parts:

1. Passing a "group" to the Reader and have it track it's offset  `__consumer_offsets`

<img width="629" alt="screen shot 2017-10-17 at 5 11 01 pm" src="https://user-images.githubusercontent.com/739782/31695242-43c11888-b35e-11e7-83fe-67aed5cef35b.png">

https://www.slideshare.net/jjkoshy/offset-management-in-kafka

2. All the functionality described by @wadearnold in original issue 🤘 

thoughts?

**Comment by achille-roussel** [comment]
This seems like a good idea to me 👍 

As we move more and more of our services to using kafka-go here at Segment we'll need support for consumer groups eventually, so unless I've received a contribution by then I will likely have to add support at some point in a not-too-far future, but I don't have a hard date yet on when that would happen.

**Comment by burdiyan** [comment]
How are you using Kafka without consumer groups? Are your services mostly stateteful, each configured handling specific partitions of specific topics? Can you give some examples of your services (apart from the deduper on your blog :))

Thanks!

**Comment by achille-roussel** [comment]
The way we've been doing it is using consul locks to negotiate which service is consuming from which partition. I a topic has 10 partitions we start 10 services and each of them attempts to acquire a lock on consul keys `<prefix>/locks/<partition>`. Then they manage they store their offset in an Aurora database. This is used to do atomic handoffs (through 3-phase commits) between the consumer handling a partition and the services that are intended to receive the messages. This mechanism ensure "exactly-once" message passing between kafka and the destinations, preventing duplicates or data loss.
Consumer groups abstract offset management in a way that would make it much harder to build a system like this one, this is why we haven't used them in this specific case, however we have other - simpler - pipelines internally that do use consumer groups, those were built before kafka-go was a thing and are still using sarama.

**Comment by burdiyan** [comment]
@achille-roussel Great! Thank's a lot for sharing! We are using consumer groups extensively, and that's why we had to stay with Sarama (which we anyway abstracted with a simple reader/writer semantics). But kafka-go looks quite interesting.

**Comment by dangkaka** [comment]
hi @achille-roussel, any update on this?

**Comment by savaki** [comment]
I think I'll take a stab a simple consumer group implementation this weekend.

**Comment by achille-roussel** [comment]
I opened https://github.com/segmentio/kafka-go/issues/44 as a place to discuss the current work that @savaki has started on consumer groups.

**Comment by achille-roussel** [comment]
For those of you who are following here, we've just merged @savaki's contribution to add support for consumer groups.

**Comment by burdiyan** [comment]
Great news! So this issue can be closed then?

**Comment by achille-roussel** [comment]
I'll close when I tag a release with the consumer group work in it, there are still a couple of implementation details to polish. But the code is good enough to be tested, and feedback are welcome!

**Comment by flowchartsman** [comment]
PR ref: https://github.com/segmentio/kafka-go/pull/55

**Comment by flowchartsman** [comment]
This PR is merged, so shouldn't this issue be closed? If it's blocked on tests, I feel like that should have a separate issue to avoid confusion.

**Comment by achille-roussel** [comment]
Yep, consumer groups have been supported for a while now, we can close this issues which was tracking the feature request.
